lighthouse-core/gather/gatherers/image-elements.js

```diff
@@ -308,30 +312,39 @@ class ImageElements extends Gatherer {
 
     /** @type {Array<LH.Artifacts.ImageElement>} */
     const imageUsage = [];
-    const top50Images = Object.values(indexedNetworkRecords)
-      .sort((a, b) => b.resourceSize - a.resourceSize)
-      .slice(0, 50);
+
     await Promise.all([
       driver.sendCommand('DOM.enable'),
       driver.sendCommand('CSS.enable'),
       driver.sendCommand('DOM.getDocument', {depth: -1, pierce: true}),
     ]);
 
+    // Sort (in-place) as largest images descending
+    elements.sort((a, b) => {
+      const aRecord = indexedNetworkRecords[a.src] || {};
+      const bRecord = indexedNetworkRecords[b.src] || {};
+      return bRecord.resourceSize - aRecord.resourceSize;
+    });
+
+    // Don't allow more than 3s of this expensive protocol work
+    let hasBudget = true;
+    setTimeout(_ => (hasBudget = false), 3000);
+
     for (let element of elements) {
       // Pull some of our information directly off the network record.
       const networkRecord = indexedNetworkRecords[element.src] || {};
       element.mimeType = networkRecord.mimeType;
 
-      if (!element.isInShadowDOM && !element.isCss) {
+      // Need source rules to determine if sized via CSS (for unsized-images).
+      if (!element.isInShadowDOM && !element.isCss && hasBudget) {
```

Comment:
Adding this bailout is the big change. Won't this lead to false failures? My read of

https://github.com/GoogleChrome/lighthouse/blob/aaf136519ce726e4a0e85cf256de110849fa5a6f/lighthouse-core/audits/unsized-images.js#L70-L82

is that if `fetchSourceRules` isn't run, then `cssWidth` and `cssHeight` will be `undefined`, and so images could be falsely marked as unsized. We should probably fail the other way, not making a claim if we don't have the info?

Reply:
Oh, I did miss this change. The previous PR was just focused on saving time for the second request for the source rules, and that's all I was looking at.

So the long tail is not that the 50 largest elements take a long time in `fetchElementWithSizeInformation `, but that all the elements take a long time to go thru `fetchSourceRules`

> We should probably fail the other way, not making a claim if we don't have the info?

Should we mark these fields `null` to signify we don't know, and `undefined` if the element size is not specified (aka when `getEffectiveSizingRule` returns undefined)? Or put `cssWidth/cssHeight` into an optional property `cssSizing?: {width?: string, height?: string}`?

Reply:
we discussed in chat..  highlights from that:

> I tend to think of `undefined` as equivalent to "not set" and `null` as equivalent to "set and has no value".

👍 👍 👍 👍 👍 👍 

idea for `imageElement.cssSizing` obj width width/height props.. obj is appropriately null/undefined if skipped etc.
obj could include the naturalWidth bits as well and be `fetchedDimensions` or we keep 'em separate.

> it would be nice on the next major to do a minor rethink of ImageElements to make it easier on audits using them, since naturalWidth/naturalHeight is in a similar but not quite the same situation as cssWidth/cssHeight after this change
> (to add to the null/undefined confusion, don't forget the return value of el.getAttribute('notASetAttribute') :)